File: csharp/ShrewdEvolver/LogNavigator.cs
// tag: 20210724T0000
using System;
using System.Collections.Generic;
using System.Text;

namespace AaronicSubstances.ShrewdEvolver
{
    public class LogNavigator<T>
    {
        private readonly IList<T> _logs;

        public LogNavigator(IList<T> logs)
        {
            _logs = logs;
        }

        public int NextIndex { get; private set;  } = 0;

        public bool HasNext()
        {
            return NextIndex < _logs.Count;
        }

        public T Next()
        {
            return _logs[NextIndex++];
        }

        public T Next(Predicate<T> searchCondition)
        {
            return Next(searchCondition, null);
        }

        public T Next(Predicate<T> searchCondition, Predicate<T> stopCondition)
        {
            if (searchCondition == null)
            {
                throw new ArgumentNullException(nameof(searchCondition));
            }
            int stopIndex = _logs.Count;
            if (stopCondition != null)
            {
                for (int i = NextIndex; i < _logs.Count; i++)
                {
                    if (stopCondition.Invoke(_logs[i]))
                    {
                        stopIndex = i;
                        break;
                    }
                }
            }
            for (int i = NextIndex; i < stopIndex; i++)
            {
                T log = _logs[i];
                if (searchCondition.Invoke(log))
                {
                    NextIndex = i + 1;
                    return log;
                }
            }
            return default;
        }
    }
}

File: java/src/main/java/com/aaronicsubstances/shrewd/evolver/LogNavigator.java
package com.aaronicsubstances.shrewd.evolver;

import java.util.List;
import java.util.Objects;
import java.util.function.Predicate;

public class LogNavigator<T> {
    private final List<T> logs;
    private int nextIndex = 0;

    public LogNavigator(List<T> logs) {
        this.logs = logs;
    }
    
    public int nextIndex() {
        return nextIndex;
    }
    
    public boolean hasNext() {
        return nextIndex < logs.size();
    }
    
    public T next() {
        return logs.get(nextIndex++);
    }
    
    public T next(Predicate<T> searchCondition) {
        return next(searchCondition, null);
    }
    
    public T next(Predicate<T> searchCondition, Predicate<T> stopCondition) {
        Objects.requireNonNull(searchCondition, "searchCondition");
        int stopIndex = logs.size();
        if (stopCondition != null) {
            for (int i = nextIndex; i < logs.size(); i++) {
                if (stopCondition.test(logs.get(i))) {
                    stopIndex = i;
                    break;
                }
            }
        }
        for (int i = nextIndex; i < stopIndex; i++) {
            T log = logs.get(i);
            if (searchCondition.test(log)) {
                nextIndex = i + 1;
                return log;
            }
        }
        return null;
    }
}